lib/aerospike/cluster.rb

```diff
@@ -229,84 +218,96 @@ def launch_tend_thread
       end
     end
 
+    # Check health of all nodes in cluster
     def tend
-      nodes = self.nodes
+      was_changed = refresh_nodes
+
+      return unless was_changed
+
+      update_cluster_features
+      notify_cluster_config_changed
+      # only log the tend finish IF the number of nodes has been changed.
+      # This prevents spamming the log on every tend interval
+      log_tend_stats(nodes)
+    end
+
+    # Refresh status of all nodes in cluster. Adds new nodes and/or removes
+    # unhealty ones
+    def refresh_nodes
       cluster_config_changed = false
 
-      # All node additions/deletions are performed in tend thread.
-      # If active nodes don't exist, seed cluster.
+      nodes = self.nodes
       if nodes.empty?
-        Aerospike.logger.info("No connections available; seeding...")
         seed_nodes
         cluster_config_changed = true
-
-        # refresh nodes list after seeding
         nodes = self.nodes
       end
 
-      # Refresh all known nodes.
-      friend_list = []
-      refresh_count = 0
+      peers = Peers.new
 
-      # Clear node reference counts.
+      # Clear node reference count
       nodes.each do |node|
-        node.reference_count.value = 0
-        node.responded.value = false
+        node.reset_reference_count!
+        node.reset_responded!
+        node.partition_generation.reset_changed!
+        # Using peers is default true
+        peers.use_peers = false unless node.supports_feature?('peers')
+      end
 
-        if node.active?
-          begin
-            friends = node.refresh
-            refresh_count += 1
-            friend_list.concat(friends) if friends
-          rescue => e
-            Aerospike.logger.error("Node `#{node}` refresh failed: #{e}")
-            Aerospike.logger.error(e.backtrace.join("\n"))
-          end
+      # refresh all known nodes
+      nodes.each do |node|
+        Node::Refresh::Info.(node, peers)
```

Comment:
Still a bit uneasy about this use of service modules. Doesn't look very object-oriented either, that you are calling a method of the Node model but then need to pass the node instance as the first parameter.

Maybe can hide this under a wrapper method in the Node class, e.g.

    class Node
      def refresh_info(peers)
        Node::Refresh::Info.(self, peers)
      end
    end

    class Cluster
      def refresh_nodes
        ...
        nodes.each do |node|
          node.refresh_info(peers)
        end
        ...
      end
    end

Then it just becomes an implementation detail of the Node model that other collaborators don't need to be aware of. Same for the other Refresh::* methods.

Reply:
I understand why you think that. Think of the idea service modules as extracting "business logic" or "business operations" and not have classes to know about the dependencies to carry out these, which tend to avoid convoluted and huge class definitions with a lot of business logic built in. Otherwise, eg. in order to perform a peer refresh the node has to eg. know about `Info` commands and a `NodeValidator`, which I don't necessarily think it should. Also, I don't see why these operations have to object-oriented necessarily; they are more a definition of steps that need to happen in order to carry out the operation.

Happy to find a time and place to discuss this more; I really enjoy thinking about these things and to learn from other experiences! 

But your suggestion is a good compromize. I'll go ahead and add that